resolver/resolver.cc

```diff
@@ -552,46 +552,61 @@ class ResolveConstantsWalk {
+
+        auto argumentNumber = 0;
+        auto encounteredError = false;
+        for (auto &arg : send->args) {
+            auto *id = ast::cast_tree<ast::ConstantLit>(arg);
+
+            if (id == nullptr || !id->symbol.exists() || !id->symbol.data(gs)->isClassOrModule()) {
+                if (auto e = gs.beginError(core::Loc(todo.file, send->loc),
+                                           core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Argument to `{}` must be statically resolvable to a module",
+                                send->fun.data(gs)->show(gs));
+                }
+                return;
             }
-            return;
-        }
-        if (id->symbol.data(gs)->isClassOrModuleClass()) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("`{}` is a class, not a module; Only modules may be mixins", id->symbol.data(gs)->show(gs));
+            if (id->symbol.data(gs)->isClassOrModuleClass()) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("`{}` is a class, not a module; Only modules may be mixins",
+                                id->symbol.data(gs)->show(gs));
+                }
+                encounteredError = true;
             }
-            return;
-        }
-        if (id->symbol == owner) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Must not pass your self to `{}`", send->fun.data(gs)->show(gs));
+            if (id->symbol == owner) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Must not pass your self to `{}`", send->fun.data(gs)->show(gs));
+                }
+                encounteredError = true;
             }
-            return;
-        }
-        auto existing = owner.data(gs)->findMember(gs, core::Names::classMethods());
-        if (existing.exists() && existing != id->symbol) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Redeclaring `{}` from module `{}` to module `{}`", send->fun.data(gs)->show(gs),
-                            existing.data(gs)->show(gs), id->symbol.data(gs)->show(gs));
+            if (encounteredError) {
+                continue;
             }
-            return;
+
+            // Get the fake property holding the mixes
+            auto mixData = id->symbol.data(gs)->findMember(gs, core::Names::mixedInClassMethods());
```

Comment:
```suggestion
            auto mixData = owner.data(gs)->findMember(gs, core::Names::mixedInClassMethods());
```
Here's your non-termination bug -- the mixData was being looked up on the wrong symbol.

Reply:
For more of an explanation about what was happening, when the thing being used with `mixes_in_class_methods` had a `mixedInClassMethods` field present, this would find it and then record an entry. When that symbol is the same as the one being processed, a loop is made in the symbol table, which is what caused the non-termination behavior in the global pass.

Reply:
Thank you very much @elliottt. Was able to find the weird singleton class behaviour but couldn't pinpoint why. Appreciate you taking the time to debug this 😊